torch/csrc/jit/codegen/cuda/root_domain_map.cpp

```diff
@@ -0,0 +1,379 @@
+#include <torch/csrc/jit/codegen/cuda/root_domain_map.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+
+#include <sstream>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+bool hasMatchingDomains(const std::vector<DomainKey>& domains,
+                        const DisjointSet<DomainKey, DomainKeyHash>& eq_set) {
+  for (const auto& key: domains) {
+    for (const auto& other_key: domains) {
+      if (key == other_key) continue;
+      const auto& root = other_key.td()->getRootDomain();
+      if (std::any_of(root.begin(), root.end(),
+                      [&](const IterDomain* id) {
+                        return eq_set.areEquivalent(key, DomainKey(other_key.td(), id));
+                      })) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool safeToJoin(const std::unordered_set<DomainKey, DomainKeyHash>& domains,
```

Comment:
Safe to join simply means there is no mapping of the iteration domains across a reduction.

Reply:
It should also detect mappings to multiple axes of a single root domain, which means it needs to be computed at those multiple loop bodies. I'll document what the function does, but `hasMatchingDomains` is the one that should find such cases.